src/main/ngapp/accio/src/app/editor/magic-wand.service.ts

```diff
@@ -251,20 +252,189 @@ export class MagicWandService {
+        // for 'mask-pixels'.
+        if (neighborColorDist > toleranceLimit) {
+          continue;
+        }
+
+        distances[neighborIndexReduced] =
+            Math.max(curPixelNode.distance, neighborColorDist);
+
+        // Continues lifespan of the loop.
+        toVisit.push({
+          distance: distances[neighborIndexReduced],
+          index: neighborIndexReduced});
+      }
+    }  // End of while loop.
+
+    return distances;
+  }
+
+  /**Finds shortest paths to every pixel from the vanilla pixel, and 
+   * adds organizes those distances into an array within a
+   * @returns {PreviewMask} previewMask object.
+   */
+  getPreviews(imgData: ImageData, xCoord: number, yCoord: number,
+      toleranceLimit: number): PreviewMask {
+    const shortestPaths: Array<number> =
+        this.getShortestPaths(imgData, xCoord, yCoord, toleranceLimit);
+    const previewMask: PreviewMask = new PreviewMask(toleranceLimit);
+
+    let tolerance: number;
+
+    for (let i = 0; i < imgData.data.length / 4; i++) {
+      // Reducing tolerance to the root reduces the number of iterations
+      // inside PreviewMask.masksAtTolerance()
+      tolerance = Math.ceil(Math.sqrt(shortestPaths[i]));
+      // Pixel index is i * 4.
+      if (previewMask.masksByTolerance[tolerance] === undefined) {
+        previewMask.masksByTolerance[tolerance] = [];
+      }
+      previewMask.masksByTolerance[tolerance].push(i * 4);
+    }
+
+    return previewMask;
+  }
 }
 
 interface Color {
- red: number,
- green: number,
- blue: number
+  red: number,
+  green: number,
+  blue: number
 }
+
+interface PixelNode {
+  distance: number,
+  index: number
+}
+
+  /**Tracks all versions of masks for 'preview' of floodfill.
+   * Versions are based on different tolerance levels.
+   **/
+export class PreviewMask {
+  masksByTolerance: Array<Array<number>> = [];
+  toleranceIndex = -1;
+  presentMask: Array<number> = [];
+
+  constructor(toleranceLimit: number) {
+    this.masksByTolerance.fill(undefined, 0, toleranceLimit + 2);
+  }
+
+  /**@returns {Set<number>} the complete floodfilled mask for
+   * the given 
+   * @param {number} tolerance level.
+   */
+  public maskAtTolerance(tolerance: number): Set<number> {
+    if (tolerance < 0) {
+      throw new Error('Tolerance input must be a non-negative number...');
+    }
+    let action: Array<number>;
+
+    while (this.toleranceIndex > tolerance) {
+      action = this.masksByTolerance[this.toleranceIndex--];
+
+      if (action === undefined) {
+        continue;
+      }
+      action.forEach(() => {
+        this.presentMask.pop();
+      });
+    }
+
+    while (this.toleranceIndex < tolerance) {
+      action = this.masksByTolerance[++this.toleranceIndex];
+
+      if (action === undefined) {
+        continue;
+      }
+      action.forEach(pixelIndex => {
+        this.presentMask.push(pixelIndex);
+      });
+    }
+
+    return new Set<number>(this.presentMask);
```

Comment:
I changed the methods for PreviewMask in a separate branch when I was trying to implement this algo on the front-end and renamed maskAtTolerance to changeMaskBy(). The reason is so that we don't end up converting to a set on every call of maskAtTolerance.
The changes were as follows:
```
  public changeMaskBy(tolerance: number): void {
    if (tolerance < 0) {
      throw new Error('Tolerance input must be a non-negative number...');
    }
    let action: Array<number>;
    while (this.toleranceIndex > tolerance) {
      action = this.masksByTolerance[this.toleranceIndex--];
      if (action === undefined) {
        continue;
      }
      action.forEach(() => {
        this.presentMask.pop();
      });
    }
    while (this.toleranceIndex < tolerance) {
      action = this.masksByTolerance[++this.toleranceIndex];
      if (action === undefined) {
        continue;
      }
      action.forEach(pixelIndex => {
        this.presentMask.push(pixelIndex);
      });
    }
  }

  public getMask(): Set<number> {
    return new Set<number>(this.presentMask);
  }
```

Reply:
I won't update it in this PR since I already made a commit on that other branch, but still wanted to let everyone know:)